docs/rfcs/01-global-object-identification/global-object-identification.md

```diff
@@ -0,0 +1,105 @@
+# Relay Object Identification
+
+## Background
+
+The following articles can be read for background on this topic:
+
+- <https://relay.dev/docs/guides/graphql-server-specification/#object-identification>
+- <https://graphql.org/learn/global-object-identification/>
+
+## Problem
+
+From <https://relay.dev/docs/guides/graphql-server-specification/>, there are two assumptions that Relay makes about a GraphQL server, in that it should provide:
+
+1. A mechanism for refetching an object.
+2. A description of how to page through connections.
+
+The second assumption is already satisfied by `@neo4j/graphql`, but a solution for the second assumption is yet to be implemented.
+
+Relay expects this to be fulfilled by a field `node` on the root query type, which returns an implementation of a `Node` interface, which contains a single field `id` of type `ID!` which implementing types must satisfy.
+
+## Solution
+
+### `Node` interface and `id` field
+
+There are two main approaches which could be taken when implementing global object identification, namely:
+
+1. Native globally unique IDs stored in the database.
+2. Synthesized globally unique IDs which are an encoded composition of the type name and a type-specific ID.
+
+The second option is significantly more attractive in that the type name can be decoded before database access and used as the label in the Cypher query.
+
+In <https://github.com/graphql/graphql-relay-js/blob/main/src/node/node.js> there is an established standard of concatenating the type name and the type-specific ID with a colon in between, and then encoding this in base64.
+
+So the main solution needed is twofold:
+
+1. How does a user opt a type in to implement the `Node` interface?
+2. How does a user indicate which field should be used as the type-specific ID?
+
+#### Implementation of the `Node` interface
+
+In <https://github.com/neo4j/graphql/pull/285>, there was talk of using the `@node` directive for this purpose. However, it now has other purposes, which means that using this would mean the directive plus an argument within, at which point it is less boilerplate to explicity add `implements Node` and the necessary `id` field to the type definition. So a `Movie` type with global object identification should be defined using:
+
+```graphql
+type Movie implements Node {
+    id: ID!
+    title: String!
+}
+```
```

Comment:
The main point of contention for me in regards to this proposal is `implements Node` versus an argument in the `@node` directive.

An example of what the latter might look like:

```graphql
type Movie @node(relay: true) {
    title: String!
}
```

Advantages of `implements Node` are mainly in its explicitness - what the user sees is what they get. This is why I opted for it in the current proposal.

Whilst an argument in the `@node` directive is maybe a bit less boilerplate, there is some "magic" in the addition of the `id` field, and it's a less familiar syntax for those already familiar with Relay.

Reply:
I'm completely on the fence about this. On the one hand, the naming collision is messy, especially if you don't live and breath schemas. The user needs to know that "Node" means two different things in different contexts -- if "implements" is used, it's a global relay id in the graphql layer context; if it's the directive "@node" then it's a mapping to the neo4j db. 

In an ideal world, you could rename the interface type to something `RelayNode` but given the strict rules of relay, that's not possible. I also don't imagine you want to rename the `@node` directive.

I like your point that `implements Node` explicitly tells you what you're getting.  My concern is that if you don't live and breath schemas, the naming collision will be very confusing unless explicitly spelled out. This could perhaps be an argument for adding a parameter to the `@node` directive -- it's the single source of truth for describing how to resolve the entity in both the db and the graphql layer. 

In terms of syntax, I think the second option is probably cleaner. Just my two cents. 

```graphql
type Documentary implements Node @node(
  label: "Movie", 
  additionalLabels: "Documentary"
) {
  //
}

type Documentary @node(
  label: "Movie",
  additionalLabels: "Documentary",
  global: true
) {
  //
} 
```

Reply:
Thanks for your thoughts here too @litewarp, it's great to still see you around and making such valuable contributions to the project! I think definitely either the second option which you have proposed, or what is currently proposed with explicit `implements Node` are our two choices. Just need to have some discussions and settle on one!

Appreciate it has taken a while to get around to this. Rough plan of action is to close #285 (probably in an unrecoverable state of merge conflicts now I imagine), but use the work done there by you as the basis for the solution we agree on here.

Reply:
@angrykoala and @danstarns, can we get some thoughts here on the following options?

@litewarp is very right in that we need to make some progress on this RFC so that it can be implemented, either by ourselves or the community.

## Option 1

Explicit implementation of the `Node` interface. `id` field must be defined by users to fulfil the interface contract.

**Pros:**
- Very clear for users who are familiar with Relay (which really is our target audience for this feature)
- No confusion about any generated `id` field, where it came from, why it is there or what it returns
- Clear separation of concerns for `Node` and `@node` (see related con)

**Cons:**
- More boilerplate
- Potential confusion of `Node` vs `@node` (see related pro)

```gql
type Movie implements Node {
    id: ID!
    title: String!
}
```

## Option 2

Adding a `global` argument to the existing `@node` directive. @litewarp, would your vision for this approach be that `id` is automatically added, or manually specified by users?

Pros:
- Minimum amount of boilerplate for implementation

Cons:
- `@node` is no longer a single purpose directive, sets a bad precedent

```gql
type Movie @node(global: true) {
    id: ID! # Would this need to be defined by user or automatically added?
    title: String!
}
```

Reply:
Personally, I think I prefer option 1. I think the only users who are going to want to turn this on are hardcore Relay users, who will be familiar with the syntax and the underlying behaviours.

Reply:
I'm also more inclined with option 1, although I'm slightly concerned about the `Node` vs `@node` naming collision